### PR TITLE
Implemented processing of orders past send time

### PR DESCRIPTION
**Pull request by acn-sbuad**
## Related Issue(s)
- #185 

## Verification
- [x] **Your** code builds clean without any errors or warnings
- [x] Manual testing done (required)
- [ ] Relevant automated test added (if you find this hard, leave it and we'll help out)
- [x] All tests run green

## Documentation
- [ ] User documentation is updated with a separate linked PR in [altinn-studio-docs.](https://github.com/Altinn/altinn-studio-docs) (if applicable)

### CODE REVIEW

File: src/Altinn.Notifications.Persistence/Migration/v0.03/01-setup-functions.sql [hunk 1]
```diff
@@ -0,0 +1,17 @@
+DROP FUNCTION IF EXISTS notifications.getorders_pastsendtime_updatestatus();
+CREATE OR REPLACE FUNCTION notifications.getorders_pastsendtime_updatestatus()
```
Comment (acn-sbuad) [inline src/Altinn.Notifications.Persistence/Migration/v0.03/01-setup-functions.sql:2]: i guess v0.02 and v0.03 can be merged to one version @tba76

File: src/Altinn.Notifications.Persistence/Repository/OrderRepository.cs [hunk 2]
```diff
@@ -1,4 +1,6 @@
-﻿using Altinn.Notifications.Core.Models.NotificationTemplate;
```
Comment (acn-sbuad) [inline src/Altinn.Notifications.Persistence/Repository/OrderRepository.cs:1]: do we still need System.Text.Json here if using the static method?

File: src/Altinn.Notifications.Persistence/Migration/v0.03/01-setup-functions.sql (hunk 1 above)
Comment (tba76) [reply src/Altinn.Notifications.Persistence/Migration/v0.03/01-setup-functions.sql]: yes

File: src/Altinn.Notifications.Persistence/Repository/OrderRepository.cs (hunk 2 above)
Comment (tba76) [reply src/Altinn.Notifications.Persistence/Repository/OrderRepository.cs]: removed

File: src/Altinn.Notifications.Core/Services/OrderProcessingService.cs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        NotificationChannel ch = order.NotificationChannel;
+        EmailTemplate? emailTemplate = order.Templates.Find(t => t.Type == NotificationTemplateType.Email) as EmailTemplate;
+
+        foreach (Recipient recipient in order.Recipients)
+        {
+            switch (ch)
+            {
+                case NotificationChannel.Email:
+
+                    await _emailService.CreateEmailNotification(order.Id, order.RequestedSendTime, recipient);
```
Comment (acn-sbuad) [inline src/Altinn.Notifications.Core/Services/OrderProcessingService.cs:59]: Error handling. What if this fails for a recipient in the list ? Should we go on processing the remaining? 
How do we ensure completed status is only set if all are successful?

File: src/Altinn.Notifications.Core/Models/Orders/NotificationOrder.cs [hunk 4]
```diff
@@ -102,4 +102,32 @@ public string Serialize()
...
+        {
+            return false;
+        }
+
+        try
+        {
+            parsedOutput = Deserialize(input!);
+
+            value = parsedOutput!;
+            return value.Id != null;
```
Comment (acn-sbuad) [inline src/Altinn.Notifications.Core/Models/Orders/NotificationOrder.cs:124]: how many properties should we check to ensure it is a valid order that has been parsed?

File: src/Altinn.Notifications.Core/Configuration/KafkaSettings.cs [hunk 5]
```diff
@@ -0,0 +1,27 @@
+﻿namespace Altinn.Notifications.Core.Configuration;
+
+/// <summary>
+/// Configuration object used to hold settings for a kafka integration settings.
```
Comment (acn-sbuad) [inline src/Altinn.Notifications.Core/Configuration/KafkaSettings.cs:4]:
```suggestion
/// Configuration object used to hold integration settings for a Kafka.
```

File: src/Altinn.Notifications.Core/Integrations/Consumers/PastDueOrdersConsumer.cs [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+using Confluent.Kafka;
+
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Altinn.Notifications.Core.Integrations.Consumers;
+
+/// <summary>
+/// Kafka Consumer class for past due orders
```
Comment (acn-sbuad) [inline src/Altinn.Notifications.Core/Integrations/Consumers/PastDueOrdersConsumer.cs:14]:
```suggestion
/// Kafka consumer class for past due orders
```

File: src/Altinn.Notifications.Core/Integrations/Consumers/PastDueOrdersConsumer.cs [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    private readonly IOrderProcessingService _orderProcessingService;
+    private readonly ILogger<PastDueOrdersConsumer> _logger;
+    private readonly KafkaSettings _settings;
+    private readonly CancellationTokenSource _cancellationTokenSource;
+    private readonly IConsumer<string, string> _consumer;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="PastDueOrdersConsumer"/> class.
+    /// </summary>
+    public PastDueOrdersConsumer(IOrderProcessingService orderProcessingService, IOptions<KafkaSettings> settings, ILogger<PastDueOrdersConsumer> logger)
```
Comment (acn-sbuad) [inline src/Altinn.Notifications.Core/Integrations/Consumers/PastDueOrdersConsumer.cs:27]:
```suggestion
    public PastDueOrdersConsumer(
    	IOrderProcessingService orderProcessingService, 
    	IOptions<KafkaSettings> settings, 
    	ILogger<PastDueOrdersConsumer> logger)
```

File: src/Altinn.Notifications.Core/Integrations/Interfaces/IKafkaProducer.cs [hunk 8]
```diff
@@ -0,0 +1,20 @@
+﻿namespace Altinn.Notifications.Core.Integrations.Interfaces;
+
+/// <summary>
+/// Interface for handeling all producer actions for kafka
```
Comment (acn-sbuad) [inline src/Altinn.Notifications.Core/Integrations/Interfaces/IKafkaProducer.cs:4]:
```suggestion
/// Interface for handling all producer actions for Kafka
```

File: src/Altinn.Notifications.Core/Integrations/Interfaces/IKafkaProducer.cs [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+/// </summary>
+public interface IKafkaProducer
+{
+    /// <summary>
+    /// Produces a message on the provided kafka topic
```
Comment (acn-sbuad) [inline src/Altinn.Notifications.Core/Integrations/Interfaces/IKafkaProducer.cs:9]:
```suggestion
    /// Produces a message on the provided Kafka topic 
```

File: test/Altinn.Notifications.Tests/AssertUtils.cs [hunk 10]
```diff
@@ -0,0 +1,14 @@
+﻿using Xunit;
+
+namespace Altinn.Notifications.Tests;
+public static class AssertUtils
+{
+    /// <summary>
+    /// Checks equality for two types
```
Comment (acn-sbuad) [inline test/Altinn.Notifications.Tests/AssertUtils.cs:7]:
```suggestion
    /// Checks equivalency of two object
```

File: src/Altinn.Notifications.Core/Repository/Interfaces/IEmailNotificationsRepository.cs [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+/// </summary>
+public interface IEmailNotificationsRepository
+{
+    /// <summary>
+    /// Adds a new email notification to the database
+    /// </summary>
+    public Task AddNotification(EmailNotification notification, DateTime expiry);
+
+    /// <summary>
+    /// Retrieves all email notifications with status 'new'
```
Comment (acn-sbuad) [inline src/Altinn.Notifications.Core/Repository/Interfaces/IEmailNotificationsRepository.cs:23]:
```suggestion
    /// Retrieves all email notifications with status 'New'
```

File: src/Altinn.Notifications.Core/Repository/Interfaces/IEmailNotificationsRepository.cs [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+    /// </summary>
+    /// <returns></returns>
```
Comment (acn-sbuad) [inline src/Altinn.Notifications.Core/Repository/Interfaces/IEmailNotificationsRepository.cs:25]:
```suggestion
```

Comment (acn-sbuad) [inline src/Altinn.Notifications.Core/Repository/Interfaces/IEmailNotificationsRepository.cs:25]:
```suggestion
    /// <returns>A list of emails</returns>
```

File: src/Altinn.Notifications.Core/Models/Orders/NotificationOrder.cs [hunk 13]
```diff
@@ -102,4 +102,32 @@ public string Serialize()
                 Converters = { new JsonStringEnumConverter() }
             });
     }
+
+    /// <summary>
+    /// Try to parse a json string into the <see cref="NotificationOrder"/>
```
Comment (acn-sbuad) [inline src/Altinn.Notifications.Core/Models/Orders/NotificationOrder.cs:107]:
```suggestion
    /// Try to parse a json string into a<see cref="NotificationOrder"/>
```

File: src/Altinn.Notifications.Core/Models/Notification/NotificationResult.cs [hunk 14]
```diff
@@ -0,0 +1,29 @@
+﻿using Altinn.Notifications.Core.Services.Interfaces;
+
+namespace Altinn.Notifications.Core.Models.Notification;
+
+/// <summary>
+/// Notification result class
```
Comment (acn-sbuad) [inline src/Altinn.Notifications.Core/Models/Notification/NotificationResult.cs:6]:
```suggestion
/// A class represednting a notification result
```

File: src/Altinn.Notifications.Core/Services/Interfaces/IEmailNotificationService.cs [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+/// </summary>
+public interface IEmailNotificationService
+{
+    /// <summary>
+    /// Process all email notifications.
+    /// </summary>
+    public Task CreateNotification(string orderId, DateTime requestedSendTime, Recipient recipient);
+
+    /// <summary>
+    /// Stats the process of sending all ready email notifications
```
Comment (acn-sbuad) [inline src/Altinn.Notifications.Core/Services/Interfaces/IEmailNotificationService.cs:16]:
```suggestion
    /// Starts the process of sending all ready email notifications
```

File: src/Altinn.Notifications.Core/Services/Interfaces/IEmailNotificationService.cs [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+    /// </summary>
+    /// <returns></returns>
```
Comment (acn-sbuad) [inline src/Altinn.Notifications.Core/Services/Interfaces/IEmailNotificationService.cs:18]:
```suggestion
```

File: src/Altinn.Notifications.Core/Services/Interfaces/IEmailNotificationService.cs [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+
+namespace Altinn.Notifications.Core.Services.Interfaces;
+
+/// <summary>
+/// Interface for email notification service
+/// </summary>
+public interface IEmailNotificationService
+{
+    /// <summary>
+    /// Process all email notifications.
```
Comment (acn-sbuad) [inline src/Altinn.Notifications.Core/Services/Interfaces/IEmailNotificationService.cs:11]:
```suggestion
    /// Creates a new email notification based on the provided orderId and recipient
```

File: src/Altinn.Notifications.Persistence/Altinn.Notifications.Persistence.csproj [hunk 18]
```diff
@@ -13,7 +13,7 @@
     <PackageReference Include="Npgsql" Version="7.0.4" />
     <PackageReference Include="Npgsql.DependencyInjection" Version="7.0.4" />
     <PackageReference Include="Yuniql.AspNetCore" Version="1.2.25" />
-    <PackageReference Include="Yuniql.PostgreSql" Version="1.3.15" />
+    <PackageReference Include="Yuniql.PostgreSql" Version="1.3.15" />
```
Comment (acn-sbuad) [inline src/Altinn.Notifications.Persistence/Altinn.Notifications.Persistence.csproj:16]:
```suggestion
    <PackageReference Include="Yuniql.PostgreSql" Version="1.3.15" />
```

File: test/Altinn.Notifications.Tests/Notifications.Persistence/TestingExtensions/WebApplicationExtensionsTests.cs [hunk 19]
```diff
@@ -0,0 +1,24 @@
...
+using Xunit;
+
+namespace Altinn.Notifications.Tests.Notifications.Persistence.TestingExtensions;
+
+public class WebApplicationExtensions
+{
+    [Fact]
+    public void SetUpPostgreSql_PostgreSettingsMissing_ThrowsException()
+    {
+
```
Comment (acn-sbuad) [inline test/Altinn.Notifications.Tests/Notifications.Persistence/TestingExtensions/WebApplicationExtensionsTests.cs:17]:
```suggestion
```

File: test/Altinn.Notifications.Tests/Notifications.Persistence/TestingExtensions/ServiceCollectionExtensionsTests.cs [hunk 20]
```diff
@@ -0,0 +1,30 @@
...
+
+using Xunit;
+
+namespace Altinn.Notifications.Tests.Notifications.Persistence.TestingExtensions;
+
+public class ServiceCollectionExtensionsTests
+{
+
+    [Fact]
+    public void AddPostgresRepositories_PostgreSettings_ThrowsException()
```
Comment (acn-sbuad) [inline test/Altinn.Notifications.Tests/Notifications.Persistence/TestingExtensions/ServiceCollectionExtensionsTests.cs:16]:
```suggestion
    public void AddPostgresRepositories_PostgreSettingsMissing_ThrowsException()
```

File: test/Altinn.Notifications.Tests/Notifications.Persistence/TestingExtensions/ServiceCollectionExtensionsTests.cs [hunk 21]
```diff
@@ -0,0 +1,30 @@
...
+    {
+        Environment.SetEnvironmentVariable("PostgreSQLSettings_EnableDBConnection", "true");
+
+        var builder = new ConfigurationBuilder()
+            .AddEnvironmentVariables();
```
Comment (acn-sbuad) [inline test/Altinn.Notifications.Tests/Notifications.Persistence/TestingExtensions/ServiceCollectionExtensionsTests.cs:21]:
```suggestion
```

**Review by tba76 — APPROVED** [review]
🍺

**Pull request merged**